tests/apollo/test_skvbc_ro_replica.py

```diff
@@ -0,0 +1,116 @@
+from math import inf
+
+from util.bft import KEY_FILE_PREFIX, with_trio, with_bft_network
+
+
+def start_replica_cmd(builddir, replica_id):
+    """
+    Return a command that starts an skvbc replica when passed to
+    subprocess.Popen.
+
+    The replica is started with a short view change timeout and with RocksDB
+    persistence enabled (-p).
+
+    Note each arguments is an element in a list.
+    """
+    statusTimerMilli = "500"
+    viewChangeTimeoutMilli = "10000"
+
+    path = os.path.join(builddir, "tests", "simpleKVBC", "TesterReplica", "skvbc_replica")
+    return [path,
+            "-k", KEY_FILE_PREFIX,
+            "-i", str(replica_id),
+            "-s", statusTimerMilli,
+            "-p"
+            ]
+
+class SkvbcReadOnlyReplicaTest(unittest.TestCase):
+
+####################################################################################################################### 
+    @with_trio
+    @with_bft_network(start_replica_cmd=start_replica_cmd, num_ro_replicas=1)
+    @verify_linearizability
+    async def test_ro_replica_start_with_delay(self, bft_network, tracker):
+        """
+        Start up N of N regular replicas.
+        Send client commands until checkpoint 1 is reached.
+        Start read-only replica.
+        Wait for State Transfer in ReadOnlyReplica to complete.
+        """
+        [bft_network.start_replica(i) for i in range(0, bft_network.config.n)]
+        with trio.fail_after(60):  # seconds
+            async with trio.open_nursery() as nursery:
+                nursery.start_soon(tracker.send_indefinite_tracked_ops)
+                while True:
+                    with trio.move_on_after(.5):  # seconds
+                        key = ['replica', 'Gauges', 'lastStableSeqNum']
+                        replica_id = 0
+                        lastStableSeqNum = await bft_network.metrics.get(replica_id, *key)
+                        
+                        if lastStableSeqNum >= 150:
+                            #enough requests
+                            print("Consensus: lastStableSeqNum:" + str(lastStableSeqNum))  
+                            nursery.cancel_scope.cancel()
+        
+        # start the read-only replica
+        ro_replica_id = bft_network.config.n
+        bft_network.start_replica(ro_replica_id)
+        with trio.fail_after(60):  # seconds
+                while True:
+                    with trio.move_on_after(.5):  # seconds
+                        key = ['replica', 'Gauges', 'lastExecutedSeqNum']
+                        lastExecutedSeqNum = await bft_network.metrics.get(ro_replica_id, *key)
+                        # success!
+                        if lastExecutedSeqNum >= 150:
+                            print("Replica " + str(ro_replica_id) + ": lastExecutedSeqNum:" + str(lastExecutedSeqNum))
+                            break
+####################################################################################################################### 
```

Comment:
nit - I don't think the "###" line separator is necessary...

Reply:
Do you prefer another character?